### PR TITLE
feat(clustering) inline version negotiation

### PR DESCRIPTION
**Pull request by Suika-Kong**
Due to refactoring, #8775 has a lot of merging conflicts. So this is a retry of that.

I've also refactored this and designed it a little differently. And I added a check before the config sync process.

fix FT-3005

### CODE REVIEW

File: kong/clustering/services/negotiation.lua [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+  return unwrapped
+end
+
+local _M = {}
+
+local function field_validate(tbl, field, typ)
+  local v = tbl
+  for _, ind in ipairs(field) do
+    if type(v) ~= "table" then
+      error("field cannot be indexed with " .. ind)
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:46]:
```suggestion
      error("field '" .. table_concat(field, ".", 1, i-1) .. "' cannot be indexed with " .. ind)
```

File: kong-2.8.0-0.rockspec [hunk 2]
```diff
@@ -71,6 +71,8 @@ build = {
     ["kong.clustering.utils"] = "kong/clustering/utils.lua",
     ["kong.clustering.compat.removed_fields"] = "kong/clustering/compat/removed_fields.lua",
     ["kong.clustering.config_helper"] = "kong/clustering/config_helper.lua",
+    ["kong.clustering.services.negotiation"] = "kong/clustering/services/negotiation.lua",
+    ["kong.clustering.services.supported"] = "kong/clustering/services/supported.lua",
```
Comment (javierguerragiraldez) [inline kong-2.8.0-0.rockspec:76]: negotiation.proto should also be added somewhere in the rockspec

File: kong/clustering/services/negotiation.lua [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+  if type(versions) ~= "table" then
+    error("invalid versions array for service " .. name)
+  end
+
+  local supported_service = supported_services[name]
+  if not supported_service then
+    return {
+      description = "unknown service.",
+    }
+  end
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:125]: nit: 
```suggestion
    return { description = "unknown service." }
  end
```

File: kong/clustering/services/negotiation.lua [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+  local service_response = negotiate_version(name, versions, supported_service)
+  return service_response
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:128]: nit:
```suggestion
  return negotiate_version(name, versions, supported_service)
```

File: kong/clustering/services/negotiation.lua [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+end
+
+local function log_negotiation_result(name, version)
+  local ok = version.version ~= nil
+  ngx_log(NOTICE, _log_prefix, "service ",
+    (ok and "accepted" or "rejected"),
+    ": \"", name, "\"",
+    (ok and ", version: " .. version.version or ""),
+    ", ", (ok and "description" or "reason"), ": ", version.description
+  )
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:138]: this is clever, but hard to read and worse, hard to search
please do:
```suggestion
  if version.version ~= nil then
    ngx_log(NOTICE, "service accepted: \"", name, "\", version: ", version.version, ", description: ", version.description )
 
  else
    ngx_log(NOTICE, "service rejected: \"", name, "\", reason: ", version.description)
  end
```

File: kong/include/kong/services/negotiation/v1/negotiation.proto [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  string type = 2;
+  string version = 3;
+  string hostname = 4;
+}
+
+// When version is undefined, service is denied.
+message ServiceVersion {
+  string version = 1;
+  string description = 2;
+}
```
Comment (javierguerragiraldez) [inline kong/include/kong/services/negotiation/v1/negotiation.proto:34]: at least put a boolean instead of overloading `version` with two different meanings.

still strongly prefer two separate arrays, since the fields actually mean different things, and could easily evolve differently.

File: kong/clustering/services/negotiation.lua [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    if not ok then
+      ngx_log(ERR, _log_prefix, result)
+      return { error_message = result }
+    end
+
+    return result
+  end)
+end
+
+-- TODO: use event to notify other workers
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:204]: or just set in a shared resource (shdict, lmdb)

File: kong/clustering/services/supported.lua [hunk 8]
```diff
@@ -0,0 +1,8 @@
+-- DP and CP shares one list.
+-- Always from newer version to older version.
```
Comment (javierguerragiraldez) [inline kong/clustering/services/supported.lua:2]:
```suggestion
-- Always order from most to least preferred version.
```
(we might consider some very new version "experimental" and put it after the "stable" one)

File: kong/clustering/services/supported.lua [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+
+return {
+    config_sync = {
+        {version = "v1", description = "The configuration synchronizing service. (JSON and Gzip)"},
+    },
+}
```
Comment (javierguerragiraldez) [inline kong/clustering/services/supported.lua:8]: nits: spaces within brackets and newline at end of file

File: kong/clustering/wrpc_control_plane.lua [hunk 10]
```diff
@@ -38,44 +38,49 @@ local ngx_CLOSE = ngx.HTTP_CLOSE
 local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
 local _log_prefix = "[wrpc-clustering] "
 
-local wrpc_config_service
-
+local init_negotiation_server = require("kong.clustering.services.negotiation").init_negotiation_server
```
Comment (javierguerragiraldez) [inline kong/clustering/wrpc_control_plane.lua:41]: move to top of the file

**Review by javierguerragiraldez — CHANGES_REQUESTED** [review]
code structure seems very good, but needs tests to be sure of corner cases, especially about data transformations and priority logic.

most of my comments are minor, except for some dubious optimizations that reduce readability or cross abstraction responsibilities.  especially about merging the accepted and rejected structures, just because they have similar fields.

File: kong/clustering/services/negotiation.lua [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+local lower = string.lower
+
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+local KONG_VERSION
+
+-- it's so annoying that protobuf does not support map to array
+local function wrap_services(service)
+  local wrapped = {}
+  for name, version in pairs(service) do
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:22]: localize `pairs`

File: kong/clustering/services/negotiation.lua [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+  return unwrapped
+end
+
+local _M = {}
+
+local function field_validate(tbl, field, typ)
+  local v = tbl
+  for _, ind in ipairs(field) do
+    if type(v) ~= "table" then
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:45]: localize `type` `error`

File: kong/clustering/services/negotiation.lua [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+local function verify_request(body)
+  for field, typ in pairs{
+    [{
+      "node",
+    }] = "object",
+    [{
+      "node", "type",
+    }] = "string",
+    [{
+      "node", "version"
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:71]: needs a `,`

File: kong/clustering/services/negotiation.lua [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+    }] = "string",
+    [{
+      "services_requested"
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:74]: needs a `,`

File: kong/clustering/services/negotiation.lua [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+    log_negotiation_result(name, negotiated_version)
+  end
+
+  return services
+end
+
+
+local function register_client(cluster_data_plane_purge_delay, id, client_node)
+  local ok, err = kong.db.clustering_data_planes:upsert({ id = id, }, {
+    last_seen = ngx.time(),
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:161]: localize `ngx.time`

File: kong/clustering/services/negotiation.lua [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+  end)
+end
+
+-- TODO: use event to notify other workers
+-- Currently we assume only worker 0 cares about wRPC services
+local negotiated_service
+local function init_negotiated_service_tab()
+  if not negotiated_service then
+    negotiated_service = {}
+  else
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:210]: a blank line before `else`

File: kong/clustering/services/negotiation.lua [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+end
+
+local negotiation_request
+
+local function get_negotiation_request()
+  if not negotiation_request then
+    negotiation_request = {
+      node = {
+        type = "KONG",
+        version = kong.version,
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:226]: should be localize `local KONG_VERSION = kong.version`

File: kong/clustering/services/negotiation.lua [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+  if not response_data then
+    return nil, err
+  end
+
+  if response_data.error_message and not response_data.node then
+    return nil, response_data.error_message
+  end
+
+  init_negotiated_service_tab()
+  for name, version in pairs(response_data.services or {}) do
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:248]: `{}` is unwise, could we define a constant empty table?

File: kong-2.8.0-0.rockspec (hunk 2 above)
Comment (Suika-Kong) [reply kong-2.8.0-0.rockspec:76]: We don't do this in the rockspec. We will add it to packing in a build script.

File: kong/include/kong/services/negotiation/v1/negotiation.proto (hunk 6 above)
Comment (Suika-Kong) [reply kong/include/kong/services/negotiation/v1/negotiation.proto:34]: I think it's intuitive: when there's no version(it being nil), then there's not a suitable version to use.

File: kong/clustering/services/negotiation.lua (hunk 7 above)
Comment (Suika-Kong) [reply kong/clustering/services/negotiation.lua:204]: The problem is `shdict` would block all workers.

File: kong/clustering/services/supported.lua (hunk 8 above)
Comment (Suika-Kong) [reply kong/clustering/services/supported.lua:2]: From the beginning, I want to keep things simple. Probably the better way for "experimental" features is to put them into a separate experimental version.

File: kong/clustering/services/negotiation.lua (hunk 17 above)
Comment (Suika-Kong) [reply kong/clustering/services/negotiation.lua:226]: Notice that `kong.version` may be initialized later than this file.

File: kong/clustering/services/negotiation.lua [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+local supported_services = require "kong.clustering.services.supported"
+local asked_services = require "kong.clustering.services.supported"
+
+local time = ngx.time
+local var = ngx.var
+local log = ngx.log
+local ERR = ngx.ERR
+local NOTICE = ngx.NOTICE
+local _log_prefix = "[wrpc-clustering] "
+local table_clear = require "table.clear"
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:13]: `require` should be move to the top of file.

File: kong/clustering/services/negotiation.lua [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+    local field_name = table_concat(field, '.')
+    error("field \"" .. field_name .. "\" must be of type " .. typ)
+  end
+end
+
+local function verify_request(body)
+  for field, typ in pairs{
+    [{
+      "node",
+    }] = "object",
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:73]: Could we change this to `"table", -- object`, then we can simplify the code above it .

File: kong/clustering/services/negotiation.lua [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+    end
+  end
+
+  return { name = name, description = "No valid version" }
+end
+
+local function negotiate_service(name, versions)
+  name = lower(name)
+
+  if type(versions) ~= "table" then
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:119]: Could we change these `if ... error() ` to `assert(...)`?

File: kong/clustering/services/negotiation.lua [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+local function init_negotiated_service_tab()
+  if not negotiated_service then
+    negotiated_service = {}
+
+  else
+    table_clear(negotiated_service)
+  end
+end
+
+local function set_negotiated_service(name, verison)
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:215]: This function is too small, I think it is not worthy.

File: kong/clustering/wrpc_data_plane.lua [hunk 23]
```diff
@@ -43,90 +50,90 @@ function _M.new(conf, cert, cert_key)
...
+    wrpc_services = wrpc_proto.new()
+    init_negotiation_client(wrpc_services)
+    init_config_service(wrpc_services)
   end
 
-  local conf = self.conf
+  return wrpc_services
+end
+
+local function communicate_body(cp)
```
Comment (chronolaw) [inline kong/clustering/wrpc_data_plane.lua:101]: `communicate_body` is a little strange, could we change it to `communicate_impl` or others?

File: kong/clustering/wrpc_data_plane.lua [hunk 24]
```diff
@@ -43,90 +50,90 @@ function _M.new(conf, cert, cert_key)
...
+  local conf = cp.conf
 
   local log_suffix = " [" .. conf.cluster_control_plane .. "]"
-  local reconnection_delay = math.random(5, 10)
 
   local c, uri, err = clustering_utils.connect_cp(
-                        "/v1/wrpc", conf, self.cert, self.cert_key,
+                        "/v1/wrpc", conf, cp.cert, cp.cert_key,
```
Comment (chronolaw) [inline kong/clustering/wrpc_data_plane.lua:107]: I think this should be `dp`, not `cp`.

File: kong/include/kong/services/negotiation/v1/negotiation.proto (hunk 6 above)
Comment (Suika-Kong) [reply kong/include/kong/services/negotiation/v1/negotiation.proto:34]: I don't have so strong opinion on whether to introduce a boolean for that, but after consideration, I still insist to use one list instead of two. The meaning of the list is obvious: the answers of CP for all services.

**Comment by chronolaw** [comment]
I think maybe after this PR, we should add some test cases to cover wrpc feature.

File: kong/clustering/services/negotiation.lua (hunk 22 above)
Comment (Suika-Kong) [reply kong/clustering/services/negotiation.lua:215]: The reason to have this function is not to reduce the code size, but keep the abstraction.

**Comment by Suika-Kong** [comment]
> I think maybe after this PR, we should add some test cases to cover wrpc feature.

Still I need to add another feature first, so we can unblock the global rate limiting.

File: kong/clustering/services/negotiation.lua (hunk 18 above)
Comment (javierguerragiraldez) [reply kong/clustering/services/negotiation.lua]: non-escaping allocations are sunk away.  in fact, the whole pairs and loop isn't even emitted, just as if it was all surrounded by an `if respose_data.services then  for..... end end`

File: kong/clustering/services/negotiation.lua (hunk 21 above)
Comment (javierguerragiraldez) [reply kong/clustering/services/negotiation.lua:119]: i also like `assert` a lot more for readability, but it does have the issue of precalculating the error.  when it's a constant it doesn't make a difference, but it's not rare that creating a good message involves some non-trivial processing, and then you have to change the code style... better to use `error()` every time you have to emit a message and leave `assert()` only for the "just in case" places where it only makes sense to die noisily.

File: kong/clustering/services/negotiation.lua (hunk 22 above)
Comment (javierguerragiraldez) [reply kong/clustering/services/negotiation.lua:215]: it's worthy to create variables and functions just to put names on things.  function calls are basically free (and invisible to the tracer)

File: kong/include/kong/services/negotiation/v1/negotiation.proto (hunk 6 above)
Comment (javierguerragiraldez) [reply kong/include/kong/services/negotiation/v1/negotiation.proto:34]: due to differences between Lua and protobuf, there's too much ambiguity between nil and empty strings.  even booleans can be somewhat deceiving (but easier to fix).  overloading a string field will fail as soon as some other module in Lua changes the configuration of the protobuf module (yes, the configs are global).

two arrays is the most robust solution

File: kong/clustering/services/negotiation.lua (hunk 7 above)
Comment (chronolaw) [reply kong/clustering/services/negotiation.lua:203]: I agree. Let's decide it after this PR.

File: kong/include/kong/services/negotiation/v1/negotiation.proto (hunk 6 above)
Comment (Suika-Kong) [reply kong/include/kong/services/negotiation/v1/negotiation.proto:34]: 2 solutions without too much thinking:
1. Since `""` is not a good version name, we treat `""` in the same way as `nil` and consider that as not supported(and it doesn't matter what configuration `pb` is in);
2. Or just a boolean to tell if service is accepted;

I think the 2 array solution is too verbose, but if it has a strong reason I'm OK with that solution.

Comment (chronolaw) [reply kong/include/kong/services/negotiation/v1/negotiation.proto:34]: I think Javier has given a good reason for this issue. and I think this needs not too much change.

File: kong/clustering/services/negotiation.lua [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+  return unwrapped
+end
+
+local _M = {}
+
+local function field_validate(tbl, field, typ)
+  local v = tbl
+  for i, ind in ipairs(field) do
+    if type(v) ~= "table" then
+      error("field '" .. table_concat(field, ".", 1, i-1) .. "' cannot be indexed with " .. ind)
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:53]: Small lint `i - 1`, using blank.

File: kong/clustering/services/negotiation.lua [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+  if not ok then
+    log(ERR, _log_prefix, "unable to update clustering data plane status: ", err)
+    return error(err)
+  end
+end
+
+function _M.init_negotiation_server(service, conf)
+  service:import("kong.services.negotiation.v1.negotiation")
+  service:set_handler("NegotiationService.NegotiateServices", function(peer, nego_req)
+    local ok, result = pcall(function()
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:177]: localize `pcall`

File: kong/clustering/wrpc_control_plane.lua [hunk 27]
```diff
@@ -38,44 +39,48 @@ local ngx_CLOSE = ngx.HTTP_CLOSE
...
+  end)
+
+  wrpc_service:set_handler("ConfigService.ReportMetadata", function(peer, data)
+    local client = cp.clients[peer.conn]
+    if client then
+      ngx_log(ngx_INFO, _log_prefix, "received initial metadata package from client: ", client.dp_id)
+      client.basic_info = data
+      client.basic_info_semaphore:post()
+    end
+    return {
```
Comment (chronolaw) [inline kong/clustering/wrpc_control_plane.lua:68]: Small optimization, this `{ ok = 'done'}` can be defined a constant table.

File: kong/clustering/wrpc_data_plane.lua [hunk 28]
```diff
@@ -43,90 +50,90 @@ function _M.new(conf, cert, cert_key)
...
+        -- count is guaranteed to not exceed 1
+        peer.config_semaphore:post()
       end
-      return { accepted = true }
-    end)
-  end
-
-  return wrpc_config_service
+    end
+    return { accepted = true }
```
Comment (chronolaw) [inline kong/clustering/wrpc_data_plane.lua:86]: Small optimization, this `{ accepted = true }` can be defined a constant table.

File: kong/clustering/wrpc_data_plane.lua [hunk 29]
```diff
@@ -43,90 +50,90 @@ function _M.new(conf, cert, cert_key)
...
 
   do
-    local resp, err = peer:call_async("ConfigService.ReportMetadata", { plugins = self.plugins_list })
+    local ok, err = negotiate(peer)
+    if not ok then
+      error(err)
+    end
+  end
+
+  do
```
Comment (chronolaw) [inline kong/clustering/wrpc_data_plane.lua:127]: Now I think this `do ... end` is unnecessary.

File: kong/clustering/wrpc_data_plane.lua [hunk 30]
```diff
@@ -223,11 +230,31 @@ function _M:communicate(premature)
   elseif perr then
     ngx_log(ngx_ERR, _log_prefix, perr, log_suffix)
   end
+end
+
+local communicate_loop
+
+function communicate(cp, reconnection_delay)
+  return ngx.timer.at(reconnection_delay or 0, communicate_loop, cp)
```
Comment (chronolaw) [inline kong/clustering/wrpc_data_plane.lua:238]: localize `ngx.timer.at`

File: kong/clustering/wrpc_data_plane.lua (hunk 29 above)
Comment (Suika-Kong) [reply kong/clustering/wrpc_data_plane.lua:127]: This is to limit the scope of variables.

File: kong/clustering/services/negotiation.lua [hunk 31]
```diff
@@ -0,0 +1,322 @@
...
+      tbl, place = accepted, accepted_n
+    else
+      rejected_n = rejected_n + 1
+      tbl, place = rejected, rejected_n
+    end
+
+    tbl[place] = {
+      name = name,
+      version = version.version,
+      message = version.description
```
Comment (chronolaw) [inline kong/clustering/services/negotiation.lua:193]: needs a `,`

File: kong/include/kong/services/negotiation/v1/negotiation.proto [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+
+message ServiceVersions {
+  repeated ServiceVersion service = 1;
```
Comment (javierguerragiraldez) [inline kong/include/kong/services/negotiation/v1/negotiation.proto:37]: the request don't have to push the description, just the versions it supports is fine.  i see the code doesn't put the description, so it wouldn't use any space, but removing it allows you to use a `repeated string`.

also, as you mentioned, protobuf doesn't allow a map to array, but an array of sub messages is much simpler both in proto definition and in the handling code:

```
message NegotiateServicesRequest {
  DPNodeDescription node = 1;
  repeated ServiceRequest services_requested = 2;
}

message ServiceRequest {
  string name = 1;
  repeated string versions = 2;
}
```

File: kong/clustering/services/negotiation.lua [hunk 33]
```diff
@@ -0,0 +1,322 @@
...
+
+local pairs = pairs
+local ipairs = ipairs
+local type = type
+local error = error
+
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+
+-- it's so annoying that protobuf does not support map to array
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:29]: this can be removed by not using a map, especially if you simplify the request (see comment in the .proto file)

File: kong/clustering/services/negotiation.lua (hunk 12 above)
Comment (javierguerragiraldez) [reply kong/clustering/services/negotiation.lua:55]: pet peeve: `type` is compiled away.  `error` is orders of magnitude more costly than the `_G` access (even more with the string/table/concat operations done here).  no real gain localizing everything in sight.

File: kong/clustering/services/negotiation.lua [hunk 34]
```diff
@@ -0,0 +1,322 @@
...
+local function negotiate_services(services_requested)
+  local services = {}
+
+  for name, versions in pairs(services_requested) do
+    if type(name) ~= "string" or type(versions) ~= "table" then
+      error("malformed service requested " .. name)
+    end
+
+    local negotiated_version = negotiate_service(name, versions)
+    services[name] = negotiated_version
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:152]: this table is always traversed, never consulted.  it's better served by an array instead of a dictionary.

File: kong/clustering/services/negotiation.lua [hunk 35]
```diff
@@ -0,0 +1,322 @@
...
+      local services_requested = unwrap_services(nego_req.services_requested)
+      verify_request(nego_req)
+
+      nego_req.node.sync_status = verify_node_compatibility(nego_req.node)
+      local services = negotiate_services(services_requested)
+      register_client(conf.cluster_data_plane_purge_delay, dp_id, nego_req.node)
+
+      local accepted, rejected = split_services(services)
+
+      local nego_result = {
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:237]: missing CP node description

File: kong/clustering/services/negotiation.lua [hunk 36]
```diff
@@ -0,0 +1,322 @@
...
+  if not response_data then
+    return nil, err
+  end
+
+  if response_data.error_message and not response_data.node then
+    return nil, response_data.error_message
+  end
+
+  init_negotiated_service_tab()
+  local serivces = merge_servcies(response_data.services_accepted, response_data.services_rejected)
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:299]: typo: merge_servcies

File: kong/clustering/services/negotiation.lua [hunk 37]
```diff
@@ -0,0 +1,322 @@
...
+local function init_negotiated_service_tab()
+  if not negotiated_service then
+    negotiated_service = {}
+
+  else
+    table_clear(negotiated_service)
+  end
+end
+
+local function set_negotiated_service(name, verison)
```
Comment (javierguerragiraldez) [inline kong/clustering/services/negotiation.lua:266]: typo: verison

**Review by javierguerragiraldez — COMMENTED** [review]
happy to see a much more robust data representation.

there's still an issue, however: while it's perfectly OK to use a `map` in protobuf, i don't see any advantage here.  I've previously mentioned that the amount of data manipulation here makes tests particularly important (because it's hard to be sure of the "shape" of the data produced at every step); now I think that changing replacing those maps with arrays would greatly simplify the code, removing most of the complex transformations.

File: kong/clustering/services/negotiation.lua (hunk 35 above)
Comment (Suika-Kong) [reply kong/clustering/services/negotiation.lua:237]: Actually no difference but I will add it.

**Review by javierguerragiraldez — APPROVED** [review]
:+1:

File: kong/clustering/services/supported.lua [hunk 38]
```diff
@@ -0,0 +1,8 @@
+-- DP and CP shares one list.
+-- Always order from most to least preferred version.
+
+return {
+    config_sync = {
```
Comment (javierguerragiraldez) [inline kong/clustering/services/supported.lua:5]: this service as defined in the specs is called `config`

File: kong/clustering/wrpc_data_plane.lua [hunk 39]
```diff
@@ -44,83 +52,83 @@ function _M.new(conf, cert, cert_key)
...
+    end
+    local resp, err = peer:call_async("ConfigService.ReportMetadata", { plugins = dp.plugins_list })
 
     -- if resp is not nil, it must be table
     if not resp or not resp.ok then
-      ngx_log(ngx_ERR, _log_prefix, "Couldn't report basic info to CP: ", resp and resp.error or err)
-      assert(ngx.timer.at(reconnection_delay, function(premature)
-        self:communicate(premature)
-      end))
+      error("Couldn't report basic info to CP: ", resp and resp.error or err)
```
Comment (javierguerragiraldez) [inline kong/clustering/wrpc_data_plane.lua:131]:
```suggestion
      error("Couldn't report basic info to CP: " .. (resp and resp.error or err))
```

**Review by fffonion — APPROVED** [review]
let's take adding tests as next priority over other wrpc features @Suika-Kong
